ios/testmanagerd/testlistener.go

```diff
@@ -0,0 +1,200 @@
+
+func (t *TestListener) testCaseDidStartForClass(testClass string, testMethod string) {
+	t.TestSuite.TestCases = append(t.TestSuite.TestCases, TestCase{
+		ClassName:  testClass,
+		MethodName: testMethod,
+	})
+}
+
+func (t *TestListener) testCaseFailedForClass(testClass string, testMethod string, message string, file string, line uint64) {
+	testCase := t.TestSuite.findTestCase(testClass, testMethod)
+	if testCase == nil {
+		log.Warn("Received failure status for an unknown test, adding it to suite")
+		t.TestSuite.TestCases = append(t.TestSuite.TestCases, TestCase{
+			ClassName:  testClass,
+			MethodName: testMethod,
+		})
+		testCase = &t.TestSuite.TestCases[len(t.TestSuite.TestCases)-1]
+	}
+
+	testCase.Status = StatusFailed
+	testCase.Err = TestError{
+		Message: message,
+		File:    file,
+		Line:    line,
+	}
+}
+
+func (t *TestListener) testCaseDidFinishForTest(testClass string, testMethod string, status string, duration float64) {
+	testCase := t.TestSuite.findTestCase(testClass, testMethod)
+	if testCase != nil {
+		// We override "failed" status for stalled tests with the value "stalled" to be able to distinguish them later
+		if testCase.Status == StatusStalled {
+			status = StatusStalled
+		}
+
+		testCase.Status = TestCaseStatus(status)
+
+		d, err := time.ParseDuration(fmt.Sprintf("%f", duration) + "s")
+		if err != nil {
+			d = 0
+			log.WithFields(log.Fields{"error": err}).Warn("Failed parsing test case duration")
+		}
+
+		testCase.Duration = d
+	}
+}
+
+func (t *TestListener) testSuiteFinished(suiteName string, date string, testCount uint64, failures uint64, skip uint64, expectedFailure uint64, unexpectedFailure uint64, uncaughtException uint64, testDuration float64, totalDuration float64) {
+	endDate, err := time.Parse(time.DateTime+" +0000", date)
+	exitIfError("Cannot parse test suite start date", err)
+
+	t.TestSuite.EndDate = endDate
+
+	d, err := time.ParseDuration(fmt.Sprintf("%f", testDuration) + "s")
+	exitIfError("Test duration cannot be parsed", err)
+	t.TestSuite.TestDuration = d
+
+	d, err = time.ParseDuration(fmt.Sprintf("%f", totalDuration) + "s")
+	exitIfError("Test duration cannot be parsed", err)
+	t.TestSuite.TotalDuration = d
+}
+
+func (t *TestListener) LogMessage(msg string) {
+	t.logWriter.Write([]byte(msg))
+}
+
+func (t *TestListener) LogDebugMessage(msg string) {
+	t.debugLogWriter.Write([]byte(msg))
+}
+
+func (t *TestListener) TestRunnerKilled() {
+	t.err = errors.New("Test runner has been explicitly killed.")
+	close(t.executionFinished)
+}
+
+func (t *TestListener) Done() <-chan struct{} {
+	return t.executionFinished
+}
+
+func (ts *TestSuite) findTestCase(className string, methodName string) *TestCase {
+	for i, _ := range ts.TestCases {
+		tc := &ts.TestCases[i]
+		if tc.ClassName == className && tc.MethodName == methodName {
+			return tc
+		}
+	}
+
+	return nil
+}
+
+func exitIfError(msg string, err error) {
+	if err != nil {
+		log.WithFields(log.Fields{"err": err}).Fatalf(msg)
+	}
+}
```

Comment:
Remove this. Again, we must not exit the application outside the cli app.

Add a testcase that provides a date that can't be parsed. You won't be able to make `go test` succeed with having this.

Reply:
Removed. 

Added a test.

When date is invalid, we use `time.Now()` which is actually really accurate since messages are real time.